net/dns/resolver/forwarder.go

```diff
@@ -883,22 +890,10 @@ func (f *forwarder) forwardWithDestChan(ctx context.Context, query packet, respo
 			metricDNSFwdErrorNoUpstream.Add(1)
 			f.logf("no upstream resolvers set, returning SERVFAIL")
 
-			if runtime.GOOS == "darwin" || runtime.GOOS == "ios" {
-				// On apple, having no upstream resolvers here is the result a race condition where
-				// we've tried a reconfig after a major link change but the system has not yet set
-				// the resolvers for the new link.  We use SystemConfiguration to query nameservers, and
-				// the timing of when that will give us the "right" answer is non-deterministic.
-				//
-				// This will typically happen on sleep-wake cycles with a Wifi interface where
-				// it takes some random amount of time (after telling us that the interface exists)
-				// for the system to configure the dns servers.
-				//
-				// Repolling the network monitor  here is a bit odd, but if we're
-				// seeing DNS queries, it's likely that the network is now fully configured, and it's
-				// an ideal time to to requery for the nameservers.
-				f.logf("injecting network monitor event to attempt to refresh the resolvers")
-				f.netMon.InjectEvent()
-			}
+			// Attempt to recompile the DNS configuration the SERVFAIL.
+			// If we are being asked to forward queries and we have no
+			// nameservers, the network is in a bad state.
+			f.servFailRecovery()
```

Comment:
This feels like kind of a late time to attempt to catch the error. Is it possible to notice that the OS returned an empty resolver list right away and start retrying then?

Reply:
Not easily.  This isn't in response to an error condition, the OS told us it doesn't have upstream resolvers when we asked (they just aren't set yet).  We don't know when to ask again - or if we need to ask again.   At this point, we have an answer to both of those questions.

A better fix here might be to integrate our netMon with a some of the higher level constructs Apple provides (NWPathMonitor perhaps) which seems to know when the interface is fully configured (and has callbacks which may help) but that'll involve a whole lot of apple-specific refactoring and a painful amount of go->swift plumbing.  I am looking at that.  Sadly, none of those APIs will give us the nameservers.

Reply:
It not ideal, but the crux of the problem is our inability to determine whether or not no resolvers is real, or an error, or a race condition.    

Ideally, we fix the timing, which may be possible by re-plumbing netMon using higher level APIs NWNetworkMonitor which generates events which seem to happen at the right times, but that's a lot of tricky C-go plumbing which requires care and feeding - and is specific to Apple - and will probably break more things int the short term than it fixes.  It's on my radar though.

Reply:
Okay, that sounds reasonable enough as a workaround for now and we can do better in future.